<img align="right" src="tf-small.png"/>

# Search from MQL

These are examples of
[MQL](https://shebanq.ancient-data.org/static/docs/MQL-Query-Guide.pdf)
queries on
[SHEBANQ](https://shebanq.ancient-data.org/hebrew/queries), 
now expressed
as Text-Fabric search templates.

For more basic examples, see
[searchTutorial](https://github.com/etcbc/text-fabric/blob/master/docs/searchTutorial.ipynb).

*Search* in Text-Fabric is a template based way of looking for structural patterns in your dataset.

In [1]:
%load_ext autoreload
%autoreload 2

In [2]:
from tf.fabric import Fabric

In [3]:
ETCBC = 'hebrew/etcbc4c'
TF = Fabric( modules=ETCBC )

This is Text-Fabric 2.2.0
Api reference : https://github.com/ETCBC/text-fabric/wiki/Api
Tutorial      : https://github.com/ETCBC/text-fabric/blob/master/docs/tutorial.ipynb
Data sources  : https://github.com/ETCBC/text-fabric-data
Data docs     : https://etcbc.github.io/text-fabric-data
Shebanq docs  : https://shebanq.ancient-data.org/text
Slack team    : https://shebanq.slack.com/signup
Questions? Ask shebanq@ancient-data.org for an invite to Slack
107 features found and 0 ignored


In [4]:
api = TF.load('''
    rela function pdp
''')
api.makeAvailableIn(globals())

  0.00s loading features ...
   |     0.21s B rela                 from /Users/dirk/github/text-fabric-data/hebrew/etcbc4c
   |     0.08s B function             from /Users/dirk/github/text-fabric-data/hebrew/etcbc4c
   |     0.13s B pdp                  from /Users/dirk/github/text-fabric-data/hebrew/etcbc4c
   |     0.00s Feature overview: 102 nodes; 4 edges; 1 configs; 7 computeds
  5.26s All features loaded/computed - for details use loadLog()


# By Oliver Glanz

[Oliver Glanz: PP with adjective followed by noun](https://shebanq.ancient-data.org/hebrew/query?version=4b&id=547)
```
select all objects where
[phrase FOCUS typ = PP
  [word sp= prep]
  [word sp=adjv]
  [word sp=subs]
]
```
64 results having 251 words.

In [5]:
query = '''
phrase typ=PP
  word sp=prep
  <: word sp=adjv
  <: word sp=subs
'''
S.study(query)
S.showPlan()
S.count(progress=1000, limit=-1)
for r in S.fetch(amount=10):
    print(S.glean(r))

  0.00s Checking search template ...
  0.00s loading features ...
   |     0.20s B sp                   from /Users/dirk/github/text-fabric-data/hebrew/etcbc4c
   |     0.22s B typ                  from /Users/dirk/github/text-fabric-data/hebrew/etcbc4c
  0.43s All additional features loaded - for details use loadLog()
  0.43s Setting up search space for 4 objects ...
  2.56s Constraining search space with 5 relations ...
  2.64s Setting up retrieval plan ...
  2.64s Ready to deliver results from 256 nodes
Iterate over S.fetch() to get the results
See S.showPlan() to interpret the results
  2.65s The results are connected to the original search template as follows:
 0     
 1 R0  phrase typ=PP
 2 R1    word sp=prep
 3 R2    <: word sp=adjv
 4 R3    <: word sp=subs
 5     
  0.00s Counting results per 1000 up to  the end of the results ...
  0.00s Done: 64 results
phrase[אֵ֖ת רַבַּ֣ת בְּנֵֽי־עַמֹּ֑ון וְ...] אֵ֖ת  רַבַּ֣ת  בְּנֵֽי־
phrase[לָכֶ֜ם יִרְאֵ֤י שְׁמִי֙ ] לָכֶ֜ם  יִרְאֵ֤י  שְׁמִ

The number of results is right. The number of words that SHEBANQ reports
is the number of words in the phrases of the result. Let us count them:

In [6]:
print(sum([len(L.d(r[0], otype='word')) for r in S.fetch()]))

251


# By Martijn Naaijer

[Martijn Naaijer: Object clauses with >CR](https://shebanq.ancient-data.org/hebrew/query?version=4b&id=997)

```
Select all objects where 

[clause rela = Objc
   [word focus first lex = '>CR']
]
```

157 results

In [7]:
query = '''
verse
    clause rela=Objc
        =: word lex=>CR
'''
S.study(query)
S.showPlan()
S.count(progress=1000, limit=-1)
for r in sorted(S.fetch(), key=lambda x: C.rank.data[x[0]-1])[0:10]:
    print(S.glean(r))

  0.00s Checking search template ...
  0.00s loading features ...
   |     0.19s B lex                  from /Users/dirk/github/text-fabric-data/hebrew/etcbc4c
  0.20s All additional features loaded - for details use loadLog()
  0.20s Setting up search space for 3 objects ...
  0.94s Constraining search space with 3 relations ...
  0.96s Setting up retrieval plan ...
  0.96s Ready to deliver results from 284 nodes
Iterate over S.fetch() to get the results
See S.showPlan() to interpret the results
  0.96s The results are connected to the original search template as follows:
 0     
 1 R0  verse
 2 R1      clause rela=Objc
 3 R2          =: word lex=>CR
 4     
  0.00s Counting results per 1000 up to  the end of the results ...
  0.00s Done: 96 results
Genesis 14:24 clause[אֲשֶׁ֣ר אָֽכְל֣וּ הַנְּעָרִ֔ים ] אֲשֶׁ֣ר 
Genesis 18:17 clause[אֲשֶׁ֖ר אֲנִ֥י עֹשֶֽׂה׃ ] אֲשֶׁ֖ר 
Genesis 24:3 clause[אֲשֶׁ֨ר לֹֽא־תִקַּ֤ח אִשָּׁה֙ לִ...] אֲשֶׁ֨ר 
Genesis 34:11 clause[אֲשֶׁ֥ר תֹּאמְר֛וּ אֵלַ֖י ] אֲשֶׁ

We have fewer cases: 96 instead of 157.
We are working on the ETCBC version 4c, and the query has been executed against 4b.
There have been coding updates that are relevant to this query, e.g. in Genesis 43:27, which is in the results
on SHEBANQ, but not here. In 4c the `rela` is `Attr`, and not `Objc`.

In [8]:
query = '''
verse book=Genesis chapter=43 verse=27
    clause
        =: word lex=>CR
'''
S.study(query)
S.showPlan()
S.count(progress=1000, limit=-1)
results = sorted(S.fetch(), key=lambda x: C.rank.data[x[0]-1])
for r in results:
    print(r[1], F.rela.v(r[1]), S.glean(r))

  0.00s Checking search template ...
  0.00s loading features ...
  0.00s All additional features loaded - for details use loadLog()
  0.01s Setting up search space for 3 objects ...
  0.75s Constraining search space with 3 relations ...
  0.84s Setting up retrieval plan ...
  0.84s Ready to deliver results from 3 nodes
Iterate over S.fetch() to get the results
See S.showPlan() to interpret the results
  0.85s The results are connected to the original search template as follows:
 0     
 1 R0  verse book=Genesis chapter=43 verse=27
 2 R1      clause
 3 R2          =: word lex=>CR
 4     
  0.00s Counting results per 1000 up to  the end of the results ...
  0.00s Done: 1 results
431688 Attr Genesis 43:27 clause[אֲשֶׁ֣ר אֲמַרְתֶּ֑ם ] אֲשֶׁ֣ר 


# By Cody Kingham

[Cody Kingham: MI Hierarchies. p.18n49. First Person Verbs in Narrative](https://shebanq.ancient-data.org/hebrew/query?version=4b&id=1050)

```
SELECT ALL OBJECTS WHERE

[book
   [clause txt = 'N'
      [word FOCUS sp = verb
        [word ps = p1
         ]
      ]
   ]
]
OR
[book
   [clause txt = '?N'
      [word FOCUS sp = verb
        [word ps = p1
         ]
      ]
   ]
]
```

273 results.

In [9]:
query = '''
book
    clause txt=N|?N
        word sp=verb ps=p1
'''
S.study(query)
S.showPlan()
S.count(progress=1000, limit=-1)
for r in sorted(S.fetch(), key=lambda x: C.rank.data[x[0]-1])[0:10]:
    print(S.glean(r))

  0.00s Checking search template ...
  0.00s loading features ...
   |     0.21s B ps                   from /Users/dirk/github/text-fabric-data/hebrew/etcbc4c
   |     0.04s B txt                  from /Users/dirk/github/text-fabric-data/hebrew/etcbc4c
  0.26s All additional features loaded - for details use loadLog()
  0.26s Setting up search space for 3 objects ...
  0.99s Constraining search space with 2 relations ...
  1.01s Setting up retrieval plan ...
  1.04s Ready to deliver results from 557 nodes
Iterate over S.fetch() to get the results
See S.showPlan() to interpret the results
  1.04s The results are connected to the original search template as follows:
 0     
 1 R0  book
 2 R1      clause txt=N|?N
 3 R2          word sp=verb ps=p1
 4     
  0.00s Counting results per 1000 up to  the end of the results ...
  0.05s Done: 273 results
 clause[וְאֶת־יְהֹושׁ֣וּעַ צִוֵּ֔יתִי בָּ...] צִוֵּ֔יתִי 
 clause[וָאֶתְחַנַּ֖ן אֶל־יְהוָ֑ה בָּ...] אֶתְחַנַּ֖ן 
 clause[וָאֶשְׁלַ֤ח מַלְאָכִים

# By Reinoud Oosting

[Reinoud Oosting: to go + object marker](https://shebanq.ancient-data.org/hebrew/query?version=4b&id=755)

```
Select all objects
where
 [clause
   [phrase function = Pred OR function = PreC
     [word FOCUS sp = verb AND vs = qal AND lex = "HLK[" ]
         ]
    ..
    [phrase FOCUS
    [word First lex = ">T"]
   ]
 ]
OR
 [clause
    [phrase FOCUS
      [word First lex = ">T" ]
    ]
..
   [phrase function = Pred OR function = PreC
     [word FOCUS sp = verb AND vs = qal AND lex = "HLK["]
   ]
 ]
 ```
 
 4 results.
 
 This is a case where we can simplify greatly because we are not hampered
 by automatic constraints on the order of the phrases.

In [10]:
query = '''
clause
  p1:phrase function=Pred|PreC
    word sp=verb vs=qal lex=HLK[
  p2:phrase
    =: word lex=>T
  p1 # p2
'''

S.study(query)
S.showPlan()
S.count(progress=1000, limit=-1)
for r in sorted(S.fetch(), key=lambda x: C.rank.data[x[0]-1])[0:10]:
    print(S.glean(r))

  0.00s Checking search template ...
  0.00s loading features ...
   |     0.21s B vs                   from /Users/dirk/github/text-fabric-data/hebrew/etcbc4c
  0.21s All additional features loaded - for details use loadLog()
  0.21s Setting up search space for 5 objects ...
  1.84s Constraining search space with 6 relations ...
  1.85s Setting up retrieval plan ...
  1.89s Ready to deliver results from 430023 nodes
Iterate over S.fetch() to get the results
See S.showPlan() to interpret the results
  1.89s The results are connected to the original search template as follows:
 0     
 1 R0  clause
 2 R1    p1:phrase function=Pred|PreC
 3 R2      word sp=verb vs=qal lex=HLK[
 4 R3    p2:phrase
 5 R4      =: word lex=>T
 6       p1 # p2
 7     
  0.00s Counting results per 1000 up to  the end of the results ...
  0.02s Done: 4 results
clause[וַנֵּ֡לֶךְ אֵ֣ת כָּל־הַ...] phrase[נֵּ֡לֶךְ ] נֵּ֡לֶךְ  phrase[אֵ֣ת כָּל־הַמִּדְבָּ֣ר הַ...] אֵ֣ת 
clause[וְאֶת־בֵּ֤ית יְהוָה֙ אֲנִ֣י ...] phrase[הֹ

# By Reinoud Oosting (ii)

[Reinoud Oosting: To establish covenant](https://shebanq.ancient-data.org/hebrew/query?version=4b&id=1485)

```
select all objects
where

 [clause
   [phrase function = Pred OR function = PreC
     [word FOCUS sp = verb AND vs = hif AND lex = "QWM[" ]
         ]
    ..
    [phrase function = Objc
    [word FOCUS lex = "BRJT/" ]
   ]
 ]
OR
 [clause
    [phrase function = Objc
      [word FOCUS lex = "BRJT/" ]
    ]
..
   [phrase function = Pred OR function = PreC
     [word FOCUS sp = verb AND vs = hif AND lex = "QWM["]
   ]
 
]
```

13 results

In [11]:
query = '''
clause
  phrase function=Pred|PreC
    word sp=verb vs=hif lex=QWM[
  phrase function=Objc
    word lex=BRJT/
'''

S.study(query)
S.showPlan()
S.count(progress=1000, limit=-1)

  0.00s Checking search template ...
  0.00s Setting up search space for 5 objects ...
  1.94s Constraining search space with 4 relations ...
  1.95s Setting up retrieval plan ...
  1.96s Ready to deliver results from 65 nodes
Iterate over S.fetch() to get the results
See S.showPlan() to interpret the results
  1.96s The results are connected to the original search template as follows:
 0     
 1 R0  clause
 2 R1    phrase function=Pred|PreC
 3 R2      word sp=verb vs=hif lex=QWM[
 4 R3    phrase function=Objc
 5 R4      word lex=BRJT/
 6     
  0.00s Counting results per 1000 up to  the end of the results ...
  0.00s Done: 13 results


In [12]:
resultsx = sorted((L.u(r[0], otype='verse')+r for r in S.fetch()), key=lambda r: sortKey(r[0]))
for r in resultsx:
    print(S.glean(r))

Genesis 6:18 clause[וַהֲקִמֹתִ֥י אֶת־בְּרִיתִ֖י אִתָּ֑ךְ ] phrase[הֲקִמֹתִ֥י ] הֲקִמֹתִ֥י  phrase[אֶת־בְּרִיתִ֖י ] בְּרִיתִ֖י 
Genesis 9:9 clause[הִנְנִ֥י מֵקִ֛ים אֶת־בְּרִיתִ֖י אִתְּכֶ֑ם ...] phrase[מֵקִ֛ים ] מֵקִ֛ים  phrase[אֶת־בְּרִיתִ֖י ] בְּרִיתִ֖י 
Genesis 9:11 clause[וַהֲקִמֹתִ֤י אֶת־בְּרִיתִי֙ אִתְּכֶ֔ם ] phrase[הֲקִמֹתִ֤י ] הֲקִמֹתִ֤י  phrase[אֶת־בְּרִיתִי֙ ] בְּרִיתִי֙ 
Genesis 17:7 clause[וַהֲקִמֹתִ֨י אֶת־בְּרִיתִ֜י בֵּינִ֣י ...] phrase[הֲקִמֹתִ֨י ] הֲקִמֹתִ֨י  phrase[אֶת־בְּרִיתִ֜י ] בְּרִיתִ֜י 
Genesis 17:19 clause[וַהֲקִמֹתִ֨י אֶת־בְּרִיתִ֥י אִתֹּ֛ו ...] phrase[הֲקִמֹתִ֨י ] הֲקִמֹתִ֨י  phrase[אֶת־בְּרִיתִ֥י ] בְּרִיתִ֥י 
Genesis 17:21 clause[וְאֶת־בְּרִיתִ֖י אָקִ֣ים אֶת־...] phrase[אָקִ֣ים ] אָקִ֣ים  phrase[אֶת־בְּרִיתִ֖י ] בְּרִיתִ֖י 
Exodus 6:4 clause[וְגַ֨ם הֲקִמֹ֤תִי אֶת־בְּרִיתִי֙ ...] phrase[הֲקִמֹ֤תִי ] הֲקִמֹ֤תִי  phrase[אֶת־בְּרִיתִי֙ ] בְּרִיתִי֙ 
Leviticus 26:9 clause[וַהֲקִימֹתִ֥י אֶת־בְּרִיתִ֖י אִתְּכֶֽם׃ ] phrase[הֲקִימֹתִ֥י ] הֲקִימֹתִ֥י  phrase[אֶת־בְּרִית

# By Reinoud Oosting (iii)

[Reinoud Oosting: To find grace in sight of](https://shebanq.ancient-data.org/hebrew/query?version=4b&id=1484)

```
select all objects
where

 [clause
   [phrase FOCUS function = Pred OR function = PreC
     [word sp = verb AND vs = qal AND lex = "MY>[" ]
         ]
    ..
    [phrase function = Objc
    [word FOCUS lex = "XN/" ]
   ]
[phrase function = Cmpl
[word FOCUS lex = "B"]
[word FOCUS lex = "<JN/"]
]
 ]
OR
 [clause
    [phrase function = Objc
      [word FOCUS lex = "XN/" ]
    ]
[phrase function = Cmpl
[word FOCUS lex = "B"]
[word FOCUS lex = "<JN/"]
..
   [phrase function = Pred OR function = PreC
     [word FOCUS sp = verb AND vs = qal AND lex = "MY>["]
   ]
 ]
]

```

38 results

In [34]:
query = '''
clause
  p1:phrase function=Pred|PreC
    word sp=verb vs=qal lex=MY>[
  p2:phrase function=Objc
    word lex=XN/
  p3:phrase function=Cmpl
    word lex=B
    <: word lex=<JN/
  p2 << p3
'''

S.study(query)
S.showPlan(details=True)
S.count(progress=1000, limit=-1)

  0.00s Checking search template ...
  0.00s Setting up search space for 8 objects ...
  3.44s Constraining search space with 9 relations ...
  3.45s Setting up retrieval plan ...
  3.46s Ready to deliver results from 16034 nodes
Iterate over S.fetch() to get the results
See S.showPlan() to interpret the results
Search with 8 objects and 9 relations
Results are instantiations of the following objects:
node  0-clause                            (    38   choices)
node  1-phrase                            (    38   choices)
node  2-word                              (    38   choices)
node  3-phrase                            (    38   choices)
node  4-word                              (    38   choices)
node  5-phrase                            (    38   choices)
node  6-word                              ( 15768   choices)
node  7-word                              (    38   choices)
Instantiations are computed along the following relations:
node                      0-clause        (    3

# By Stephen Ku

[Stephen Ku: Verbless Clauses](https://shebanq.ancient-data.org/hebrew/query?version=4&id=1314)

```
SELECT ALL OBJECTS WHERE 

[clause  
 [phrase function IN (Subj) 
    [phrase_atom NOT rela IN (Appo,Para,Spec)
      [word FOCUS pdp IN (subs,nmpr,prps,prde,prin,adjv)
      ]
    ]
  ]
 NOTEXIST [phrase function IN (Pred)]
 ..
 NOTEXIST [phrase function IN (Pred)]
 [phrase function IN (PreC)
     NOTEXIST [word pdp IN (prep)]
     [word FOCUS pdp IN (subs,nmpr,prin,adjv) AND ls IN (card,ordn)]
 ]
]
```

1441 results with 1244 words in those results.

We do not have the `NOTEXIST` operator, and we cannot say `NOT rela IN`,
so we are at a disadvantage here.
Let's see what we can do.
We can use additional processing to furnish the template and weed out results.

The first thing is: we have to fetch all possible values of the `rela` feature,
in order to see what other values than `Appo`, `Para`, `Spec` it can take.

The function `freqList()` gives us a frequency list of values, we only need the values
other than the indicated ones, separated by a `|`.

We also need to consult the relation legend to pick the proper ordering between the
two phrases.

In [11]:
excludedRela = {'Appo', 'Para', 'Spec'}
'|'.join(x[0] for x in F.rela.freqList() if x[0] not in excludedRela)

'NA|rec|par|Adju|Attr|adj|Coor|atr|dem|Resu|Objc|Link|mod|Subj|RgRc|ReVo|Cmpl|PrAd|PreC|Sfxs'

In [12]:
print(S.relationLegend)

                      = left equal to right (as node)
                      # left unequal to right (as node)
                      < left before right (in canonical node ordering)
                      > left after right (in canonical node ordering)
                     == left occupies same slots as right
                     && left has overlapping slots with right
                     ## left and right do not have the same slot set
                     || left and right do not have common slots
                     [[ left embeds right
                     ]] left embedded in right
                     << left completely before right
                     >> left completely after right
                     =: left and right start at the same slot
                     := left and right end at the same slot
                     :: left and right start and end at the same slot
                     <: left immediately before right
                     :> left immediately after right
   

In [13]:
query = '''
clause
  p1:phrase function=Subj
    phrase_atom rela=NA|rec|par|Adju|Attr|adj|Coor|atr|dem|Resu|Objc|Link|mod|Subj|RgRc|ReVo|Cmpl|PrAd|PreC|Sfxs
      word pdp=subs|nmpr|prps|prde|prin|adjv
  p2:phrase function=PreC
    word pdp=subs|nmpr|prin|adjv ls=card|ordn

  p1 << p2
'''

S.study(query)
S.showPlan()
S.count(progress=1000, limit=-1)
for r in sorted(S.fetch(), key=lambda x: C.rank.data[x[0]-1])[0:10]:
    print(S.glean(r))

  0.00s Checking search template ...
  0.00s loading features ...
   |     0.14s B ls                   from /Users/dirk/github/text-fabric-data/hebrew/etcbc4c
  0.15s All additional features loaded - for details use loadLog()
  0.15s Setting up search space for 6 objects ...
  2.55s Constraining search space with 6 relations ...
  2.56s Setting up retrieval plan ...
  2.58s Ready to deliver results from 573297 nodes
Iterate over S.fetch() to get the results
See S.showPlan() to interpret the results
  2.58s The results are connected to the original search template as follows:
 0     
 1 R0  clause
 2 R1    p1:phrase function=Subj
 3 R2      phrase_atom rela=NA|rec|par|Adju|Attr|adj|Coor|atr|dem|Resu|Objc|Link|mod|Subj|RgRc|ReVo|Cmpl|PrAd|PreC|Sfxs
 4 R3        word pdp=subs|nmpr|prps|prde|prin|adjv
 5 R4    p2:phrase function=PreC
 6 R5      word pdp=subs|nmpr|prin|adjv ls=card|ordn
 7     
 8       p1 << p2
 9     
  0.00s Counting results per 1000 up to  the end of the results ...
  

We have too many results, because we have not posed the restrictions by the `NOTEXIST` operator.
Let's weed out the results that do not satisfy those criteria.
That is, essentially, throwing away those clauses 

* that have a phrase with `function=Pred` after the phrase with `function=Pred`
* where the second phrase has a preposition

In [14]:
indent(reset=True)
properResults = []
resultWords = set()

for r in S.fetch():
    clause = r[0]
    phrase1 = r[1]
    phrase2 = r[4]
    word1 = r[3]
    word2 = r[5]
    phrases = [p for p in L.d(clause, otype='phrase') if sortKey(p) > sortKey(phrase1)]
        
    words2 = L.d(phrase2, otype='word')
    if any(F.function.v(phrase) == 'Pred' for phrase in phrases): continue
    if any(F.pdp.v(word) == 'prep' for word in words2): continue
    resultWords |= {word1, word2}
    properResults.append(r)

info('Found {} proper results with {} words in it'.format(len(properResults), len(resultWords)))

  0.17s Found 2307 proper results with 2133 words in it


We have still many more results than the MQL query on SHEBANQ.

Let us have a look at some results words and compare them with the result words on SHEBANQ.
It is handy to fetch from SHEBANQ the csv file with query results.

In [15]:
resultsx = sorted((L.u(r[0], otype='verse')+r for r in properResults), key=lambda r: sortKey(r[0]))
resultWordsx = [(L.u(w, otype='verse')[0], w) for w in sortNodes(resultWords)]
for r in resultWordsx[0:30]:
    print(S.glean(r))

Genesis 5:4 יְמֵי־
Genesis 5:4 אָדָ֗ם 
Genesis 5:4 שְׁמֹנֶ֥ה 
Genesis 5:4 מֵאֹ֖ת 
Genesis 5:5 כָּל־
Genesis 5:5 יְמֵ֤י 
Genesis 5:5 אָדָם֙ 
Genesis 5:5 תְּשַׁ֤ע 
Genesis 5:5 מֵאֹות֙ 
Genesis 5:5 שְׁלֹשִׁ֖ים 
Genesis 5:8 כָּל־
Genesis 5:8 יְמֵי־
Genesis 5:8 שֵׁ֔ת 
Genesis 5:8 שְׁתֵּ֤ים 
Genesis 5:8 עֶשְׂרֵה֙ 
Genesis 5:8 תְשַׁ֥ע 
Genesis 5:8 מֵאֹ֖ות 
Genesis 5:11 כָּל־
Genesis 5:11 יְמֵ֣י 
Genesis 5:11 אֱנֹ֔ושׁ 
Genesis 5:11 חָמֵ֣שׁ 
Genesis 5:11 תְשַׁ֥ע 
Genesis 5:11 מֵאֹ֖ות 
Genesis 5:14 כָּל־
Genesis 5:14 יְמֵ֣י 
Genesis 5:14 קֵינָ֔ן 
Genesis 5:14 עֶ֣שֶׂר 
Genesis 5:14 תְשַׁ֥ע 
Genesis 5:14 מֵאֹ֖ות 
Genesis 5:17 כָּל־


In the list from SHEBANQ we see this:

The first thing we miss in the SHEBANQ output is

```
Genesis 5:14 עֶ֣שֶׂר
```

and in SHEBANQ we see that this word has not been marked with `ls=card|ordn`,
while in the newer ETCBC4c it is!

I have conducted a SHEBANQ query for numerals here
[Dirk Roorda: numerals](https://shebanq.ancient-data.org/hebrew/query?id=1487),
in versions 4 and 4b, 
and quite something happened with the encoding of numerals between those versions.

Let us also find the numerals in 4c:

In [16]:
S.study('''
word ls=card|ordn
''')

  0.00s Checking search template ...
  0.00s Setting up search space for 1 objects ...
  0.61s Constraining search space with 0 relations ...
  0.61s Setting up retrieval plan ...
  0.61s Ready to deliver results from 7013 nodes
Iterate over S.fetch() to get the results
See S.showPlan() to interpret the results


So we have for the amount of numerals in the ETCBC versions:

4|4b|4c
---|---|---
6839|7014|7013

On the basis of these numbers, this cannot be the sole cause of the discrepancy.

# By Dirk Roorda

[Dirk Roorda: Yesh](https://shebanq.ancient-data.org/hebrew/query?version=4b&id=556)

```
select all objects where
[book [chapter [verse
[clause
    [clause_atom
        [phrase
            [phrase_atom
                [word focus lex="JC/" OR lex=">JN/"]
            ]
        ]
    ]
]
]]]
```

926 results

In [17]:
query = '''
verse
  clause
    clause_atom
      phrase
        phrase_atom
          word lex=JC/|>JN/
'''

S.study(query)
S.showPlan()
S.count(progress=1000, limit=-1)
for r in sorted(S.fetch(), key=lambda x: C.rank.data[x[0]-1])[0:10]:
    print(S.glean(r))

  0.00s Checking search template ...
  0.00s Setting up search space for 6 objects ...
  0.77s Constraining search space with 5 relations ...
  0.81s Setting up retrieval plan ...
  0.82s Ready to deliver results from 5415 nodes
Iterate over S.fetch() to get the results
See S.showPlan() to interpret the results
  0.82s The results are connected to the original search template as follows:
 0     
 1 R0  verse
 2 R1    clause
 3 R2      clause_atom
 4 R3        phrase
 5 R4          phrase_atom
 6 R5            word lex=JC/|>JN/
 7     
  0.00s Counting results per 1000 up to  the end of the results ...
  0.02s Done: 926 results
Genesis 2:5 clause[וְאָדָ֣ם אַ֔יִן ] clause_atom[וְאָדָ֣ם אַ֔יִן ] phrase[אַ֔יִן ] phrase_atom[אַ֔יִן ] אַ֔יִן 
Genesis 5:24 clause[וְאֵינֶ֕נּוּ ] clause_atom[וְאֵינֶ֕נּוּ ] phrase[אֵינֶ֕נּוּ ] phrase_atom[אֵינֶ֕נּוּ ] אֵינֶ֕נּוּ 
Genesis 7:8 clause[אֲשֶׁ֥ר אֵינֶ֖נָּה טְהֹרָ֑ה ] clause_atom[אֲשֶׁ֥ר אֵינֶ֖נָּה טְהֹרָ֑ה ] phrase[אֵינֶ֖נָּה ] phrase_atom[אֵינֶ֖נָּה 